### PR TITLE
[Feature] Flip videos horizontally

### PR DESCRIPTION
**Pull request by YasheshSavani**
Description:

This PR will add a feature to flip videos horizontally (Left -> Right) vice versa in the sense studio at the class level. This will help users create more data and get more variability in data.

Features:
1. Users can save the flipped videos in the same class videos folder or they can also create a new class by updating the class field.
2. Users can also copy the original tags associated with the original videos to their flipped version by selecting them on the UI.

### CODE REVIEW

File: tools/sense_studio/sense_studio.py [hunk 1]
```diff
@@ -44,6 +48,9 @@
 
 socketio.init_app(app)
 
+# Training script expects videos in MP4 format
+VIDEO_EXT = '.mp4'
```
Comment (guillaumebrg) [inline tools/sense_studio/sense_studio.py:52]: It looks like `VIDEO_EXT` could be used in more places... See other comments below:

File: tools/sense_studio/sense_studio.py [hunk 2]
```diff
@@ -205,6 +212,7 @@ def project_details(project):
             stats[class_name][split] = {
                 'total': len(os.listdir(videos_dir)),
                 'tagged': len(os.listdir(tags_dir)) if os.path.exists(tags_dir) else 0,
+                'videos': natsorted([video for video in os.listdir(videos_dir) if video.endswith('.mp4')], alg=ns.IC),
```
Comment (guillaumebrg) [inline tools/sense_studio/sense_studio.py:215]: here

File: tools/sense_studio/sense_studio.py [hunk 3]
```diff
@@ -391,6 +399,85 @@ def set_timer_default():
...
+        videos_path_in = os.path.join(path, f'videos_{split}', original_class_name)
+        videos_path_out = os.path.join(path, f'videos_{split}', counterpart_class_name)
+        original_tags_path = os.path.join(path, f'tags_{split}', original_class_name)
+        counterpart_tags_path = os.path.join(path, f'tags_{split}', counterpart_class_name)
+
+        # Create directory to save flipped videos
+        os.makedirs(videos_path_out, exist_ok=True)
+        os.makedirs(counterpart_tags_path, exist_ok=True)
+
+        video_list = [video for video in os.listdir(videos_path_in) if video.endswith('.mp4')]
```
Comment (guillaumebrg) [inline tools/sense_studio/sense_studio.py:432]: here

File: tools/sense_studio/sense_studio.py [hunk 4]
```diff
@@ -391,6 +399,85 @@ def set_timer_default():
...
+                # Get flipped video output
+                flipped_video_output = ffmpeg.output(flipped_video,
+                                                     filename=os.path.join(videos_path_out, flipped_video_name))
+                # Run to render and save video
+                ffmpeg.run(flipped_video_output)
+
+                # Copy tags of original video to flipped video (in train set)
+                if video in copy_train_video_tags and split == 'train':
+                    original_tags_file = os.path.join(original_tags_path, video.replace('.mp4', '.json'))
+                    flipped_tags_file = os.path.join(counterpart_tags_path, flipped_video_name.replace('.mp4', '.json'))
```
Comment (guillaumebrg) [inline tools/sense_studio/sense_studio.py:456]: here

File: tools/sense_studio/sense_studio.py [hunk 5]
```diff
@@ -391,6 +399,85 @@ def set_timer_default():
...
+                        with open(original_tags_file) as f:
+                            original_video_tags = json.load(f)
+                        original_video_tags['file'] = flipped_video_name
+                        with open(flipped_tags_file, 'w') as f:
+                            f.write(json.dumps(original_video_tags, indent=2))
+
+                # Copy tags of original video to flipped video (in valid set)
+                if video in copy_valid_video_tags and split == 'valid':
+                    original_tags_file = os.path.join(original_tags_path, video.replace('.mp4', '.json'))
+                    flipped_tags_file = os.path.join(counterpart_tags_path, flipped_video_name.replace('.mp4', '.json'))
```
Comment (guillaumebrg) [inline tools/sense_studio/sense_studio.py:468]: here

File: tools/sense_studio/sense_studio.py [hunk 6]
```diff
@@ -391,6 +399,85 @@ def set_timer_default():
...
+        counterpart_tags_path = os.path.join(path, f'tags_{split}', counterpart_class_name)
+
+        # Create directory to save flipped videos
+        os.makedirs(videos_path_out, exist_ok=True)
+        os.makedirs(counterpart_tags_path, exist_ok=True)
+
+        video_list = [video for video in os.listdir(videos_path_in) if video.endswith('.mp4')]
+
+        for video in video_list:
+            output_video_list = [op_video for op_video in os.listdir(videos_path_out) if op_video.endswith('.mp4')]
```
Comment (guillaumebrg) [inline tools/sense_studio/sense_studio.py:435]: here

File: tools/sense_studio/sense_studio.py [hunk 7]
```diff
@@ -391,6 +399,85 @@ def set_timer_default():
...
+                if video in copy_valid_video_tags and split == 'valid':
+                    original_tags_file = os.path.join(original_tags_path, video.replace('.mp4', '.json'))
+                    flipped_tags_file = os.path.join(counterpart_tags_path, flipped_video_name.replace('.mp4', '.json'))
+
+                    if os.path.exists(original_tags_file):
+                        with open(original_tags_file) as f:
+                            original_video_tags = json.load(f)
+                        original_video_tags['file'] = flipped_video_name
+                        with open(flipped_tags_file, 'w') as f:
+                            f.write(json.dumps(original_video_tags, indent=2))
```
Comment (guillaumebrg) [inline tools/sense_studio/sense_studio.py:475]: Is there a difference between the "train" and "valid" cases? They look identical to me. Might be missing something

**Review by guillaumebrg — COMMENTED** [review]
A few comments

File: tools/sense_studio/sense_studio.py (hunk 7 above)
Comment (YasheshSavani) [reply tools/sense_studio/sense_studio.py:475]: There are two lists coming from the UI separately for train and valid, as there might be a case where both may have same named video

Comment (YasheshSavani) [reply tools/sense_studio/sense_studio.py:475]: But I'll think something if I can merge this

**Review by guillaumebrg — COMMENTED** [review]
Seems to work well. @sunny-panchal do you want to have a look or should I merge?

**Comment by sunny-panchal** [comment]
Had a quick look. You can go ahead and merge it 👍

**Review by guillaumebrg — APPROVED** [review]

**Pull request merged**